src/index.ts

```diff
@@ -1,166 +1,195 @@
+      if (oType.properties?.length > 0) {
+        return refnamespace + cppIdentifer(oType.name)
+      } else {
+        // untyped object
+        return "jsoncons::json"
       }
-      return "int64_t";
-    case "integer":
-      return "int32_t";
-    case "boolean":
-      return "bool";
-    case "object":
-      return "jsoncons::json";
-    case "array":
-      return 'std::vector<' + toCppType(property.items as Property, refnamespace) + '>';
-    case "buffer":
-      return "std::vector<uint8_t>";
+    case 'enum':
+      return refnamespace + cppIdentifer((type as EnumType).name)
+    case 'map':
+      const { keyType, valueType } = type as MapType
+      return 'std::unordered_map<' + v2ToCppTypeX(keyType, refnamespace) + ', ' + v2ToCppTypeX(valueType, refnamespace) + '>'
     default:
-      throw new Error("Can't convert property to C++ type: " + property.type);
+      throw new Error("Can't convert XTP type to C++ type: " + JSON.stringify(type))
   }
 }
 
-function toCppType(property: PropertyLike, refnamespace: string) {
-  const innerType = toCppTypeInner(property, refnamespace);
-  if (property.nullable) {
+function v2ToCppTypeX(type: XtpNormalizedType, refnamespace: string) {
+  const innerType = v2ToCppTypeXInner(type, refnamespace);
+  if (type.nullable) {
     return 'std::optional<' + innerType + '>';
   }
   return innerType;
 }
 
-function isLargeType(property: PropertyLike) {
-  return getEstimatedSize(property) > 128;
+function v2ToCppType(property: XtpTyped, refnamespace: string, required?: boolean): string {
+  const t = v2ToCppTypeX(property.xtpType, refnamespace)
+
+  // if required is unset, just return what we get back
+  if (required === undefined) return t
+
+  // if it's set and true, just return what we get back
+  if (required) return t
+
+  // otherwise it's false, so let's ensure it's optional
+  if (t.startsWith('std::optional<')) return t
+  return `std::optional<${t}>`
 }
 
-function toStorageType(param: Parameter, refnamespace: string) {
-  if (isLargeType(param)) {
-    return 'std::unique_ptr<' + toCppTypeInner(param, refnamespace) + '>';
+function v2GetEstimatedSize(type: XtpNormalizedType): number {
+  switch (type.kind) {
+    case 'string':
+      return 16
+    case 'int32':
+      return 4
+    case 'int64':
+      return 8
+    case 'float':
+      return 4
+    case 'double':
+      return 8
+    case 'byte':
+      return 1
+    case 'date-time':
+      return 14
+    case 'boolean':
+      return 1
+    case 'array':
+      return 16
+    case 'buffer':
+      return 16
+    case 'object':
+      const oType = (type as ObjectType)
+      if (oType.properties?.length > 0) {
+        return oType.properties.reduce((accumulator: number, currentValue: XtpNormalizedType) => {
+          return accumulator + v2GetEstimatedSize(currentValue);
+        }, 0);
+      } else {
+        // untyped object
+        return 64
+      }
+    case 'enum':
+      return 5
```

Comment:
1 byte discriminant + 4 byte payload?

Reply:
I think enums are the same size as `int`, but used `5` as a hack to order `int`s first, trying to sort higher level types first in structs if size is identical. `date-time` is listed as 14 for similar reasons, wanting to order it after strings as a more optimized smaller implementation should be possible in the future. Many of the numbers here are guesses.